packages/roosterjs-editor-plugins/lib/PasteManager/PasteManager.ts

```diff
@@ -1,50 +1,64 @@
-import ClipBoardData from './ClipboardData';
 import {
+    ClipBoardData,
     ContentChangedEvent,
-    ContentPosition,
     PluginDomEvent,
     PluginEvent,
     PluginEventType,
 } from 'roosterjs-editor-types';
-import { Editor, EditorPlugin } from 'roosterjs-editor-core';
-import { processImages } from './PasteUtility';
-import { fromHtml } from 'roosterjs-editor-dom';
-import { insertImage } from 'roosterjs-editor-api';
+import { Editor, EditorPlugin, buildSnapshot, restoreSnapshot } from 'roosterjs-editor-core';
+import { buildClipBoardData, insertImage } from 'roosterjs-editor-api';
 import convertPastedContentFromWord from './wordConverter/convertPastedContentFromWord';
+import removeUnsafeTags from './removeUnsafeTags';
 
-const INLINE_POSITION_STYLE = /(<\w+[^>]*style=['"][^>]*)position:[^>;'"]*/gi;
-const TEXT_WITH_BR_ONLY = /^[^<]*(<br>[^<]*)+$/i;
-const CONTAINER_HTML =
-    '<div contenteditable style="width: 1px; height: 1px; overflow: hidden; position: fixed; top: 0; left; 0; -webkit-user-select: text"></div>';
+/**
+ * Paste option
+ */
+export const enum PasteOption {
+    /**
+     * Paste with original format when copy
+     */
+    PasteAsIs = 0,
+
+    /**
+     * Paste as plain text, remove any format
+     */
+    PasteText = 1,
 
-interface WindowForIE extends Window {
-    clipboardData: DataTransfer;
+    /**
+     * Paste using current format
+     */
+    PasteWithCurrentFormat = 2,
 }
 
 /**
  * PasteManager plugin, handles onPaste event and paste content into editor
  */
 export default class PasteManager implements EditorPlugin {
     private editor: Editor;
-    private pasteContainer: HTMLElement = null;
 
     /**
      * Create an instance of PasteManager
      * @param pasteHandler An optional pasteHandler to perform extra actions after pasting.
      * Default behavior is to paste image (if any) as BASE64 inline image.
+     * @param useDirectPaste: This is a test parameter and may be removed in the future.
+     * When set to true, we retrieve HTML from clipboard directly rather than using a hidden pasting DIV,
+     * then filter out unsafe HTML tags and attributes. Although we removed some unsafe tags such as SCRIPT,
+     * OBJECT, ... But there is still risk to have other kinds of XSS scripts embeded. So please do NOT use
+     * this parameter if you don't have other XSS detecting logic outside the edtior.
+     * @param defaultPasteOption: Default option of pasting. Default value is PasteAsIs
      */
-    constructor(private readonly pasteHandler?: (clipboardData: ClipBoardData) => void) {}
+    constructor(
+        private readonly imageHandler?: (clipboardData: ClipBoardData) => void,
+        private useDirectPaste?: boolean,
```

Comment:
Shall we move 'useDirectPaste' first since we have defaultImageHandler, that way user doesn't need to explicitly pass 'null' for it.

Reply:
This is to be compatible with old code. If I change original paramater's meaning, existing code will be broken.